File: tests/test_orchestration.py
from pathlib import Path

from doc_analyse.classifiers import ClassificationResult
from doc_analyse.detection import DetectionFinding, YaraDetector
from doc_analyse.detection.detect import (
    CheapRouter,
    DECISION_HOLD,
    DECISION_REVIEW,
    DECISION_SAFE,
    YaraEvidence,
)
from doc_analyse.ingestion import TextChunker
from doc_analyse.ingestion.models import TextChunk
from doc_analyse.orchestration import (
    DocumentOrchestrator,
    analyze_document_path,
)
from doc_analyse.workers import WorkerResult


class FakeWorkerPool:
    def __init__(self):
        self.calls = []
        self.closed = False

    def classify_chunks(self, chunks):
        chunk_list = tuple(chunks)
        self.calls.append(chunk_list)

        results = []
        for chunk in chunk_list:
            verdict = "unsafe" if "RISKY-INSTRUCTION" in chunk.text else "safe"
            results.append(
                WorkerResult(
                    chunk=chunk,
                    classification=ClassificationResult(
                        verdict=verdict,
                        confidence=0.9,
                        reasons=("worker",),
                    ),
                )
            )

        return tuple(results)

    def close(self):
        self.closed = True


class FakeYara:
    """YaraDetector stand-in that flags only RISKY-INSTRUCTION."""

    def __init__(self, findings=None):
        self._findings = findings or ()

    def detect(self, chunk: TextChunk):
        if "RISKY-INSTRUCTION" not in chunk.text:
            return ()

        idx = chunk.text.index("RISKY-INSTRUCTION")
        return (
            DetectionFinding(
                span="RISKY-INSTRUCTION",
                category="instruction_override",
                severity="high",
                reason="Risk marker for test.",
                rule_id="risk_marker",
                start_char=chunk.start_char + idx,
                end_char=chunk.start_char + idx + 17,
                source=chunk.source,
                requires_llm_validation=True,
            ),
        )


class FakeRouter:
    """Router that returns HOLD for RISKY-INSTRUCTION, SAFE otherwise."""

    def __init__(self, hold_on_risk=True):
        self.hold_on_risk = hold_on_risk

    def route(self, yara_findings, pg_score):
        from doc_analyse.detection.detect import CheapChunkDecision

        if yara_findings and self.hold_on_risk:
            return CheapChunkDecision(
                decision=DECISION_HOLD,
                risk_score=50.0,
                pg_score=0.0,
                yara_score=40.0,
                findings=(),
                reason="RISKY-INSTRUCTION found.",
            )
        elif yara_findings and not self.hold_on_risk:
            return CheapChunkDecision(
                decision=DECISION_REVIEW,
                risk_score=25.0,
                pg_score=0.0,
                yara_score=20.0,
                findings=(),
                reason="RISKY-INSTRUCTION found.",
            )
        return CheapChunkDecision(
            decision=DECISION_SAFE,
            risk_score=0.0,
            pg_score=0.0,
            yara_score=0.0,
            findings=(),
            reason="No risk signals.",
        )


def test_orchestrator_end_to_end_with_index_traceability(tmp_path: Path):
    path = tmp_path / "doc.txt"
    path.write_text(
        "Normal section.\n"
        "RISKY-INSTRUCTION appears in this chunk and should route to llm.\n"
        "Normal ending.",
        encoding="utf-8",
    )
    chunker = TextChunker(chunk_size=60, chunk_overlap=0)
    pool = FakeWorkerPool()
    orchestrator = DocumentOrchestrator(
        yara=FakeYara(),
        pg=None,
        router=FakeRouter(hold_on_risk=True),
        worker_pool=pool,
    )

    result = orchestrator.analyze_path(path, chunker=chunker)

    assert result.verdict == "unsafe"
    assert len(pool.calls) == 1
    routed_chunks = pool.calls[0]
    assert len(routed_chunks) == 1
    routed_index = routed_chunks[0].metadata["chunk_index"]
    routed_result = result.chunk_result(routed_index)
    assert routed_result.routed_to_llm is True
    assert routed_result.llm_classification is not None
    assert routed_result.llm_classification.verdict == "unsafe"


def test_orchestrator_skips_worker_pool_when_no_cheap_findings(tmp_path: Path):
    path = tmp_path / "safe.txt"
    path.write_text("No risk markers in this document.", encoding="utf-8")
    pool = FakeWorkerPool()
    orchestrator = DocumentOrchestrator(
        yara=FakeYara(),
        pg=None,
        router=FakeRouter(hold_on_risk=True),
        worker_pool=pool,
    )

    result = orchestrator.analyze_path(path)

    assert result.verdict == "safe"
    assert pool.calls == []
    assert all(not cr.routed_to_llm for cr in result.chunk_results)


def test_orchestrator_routes_review_chunks_to_layer2(tmp_path: Path):
    path = tmp_path / "doc.txt"
    path.write_text("RISKY-INSTRUCTION", encoding="utf-8")
    pool = FakeWorkerPool()
    orchestrator = DocumentOrchestrator(
        yara=FakeYara(),
        pg=None,
        router=FakeRouter(hold_on_risk=False),  # REVIEW, not HOLD
        worker_pool=pool,
    )

    result = orchestrator.analyze_path(path)

    # REVIEW still goes to Layer 2
    assert len(pool.calls) == 1
    assert result.chunk_results[0].routed_to_llm is True


def test_orchestrator_context_manager_closes_worker_pool(tmp_path: Path):
    path = tmp_path / "safe.txt"
    path.write_text("safe", encoding="utf-8")
    pool = FakeWorkerPool()
    pool.closed = False

    with DocumentOrchestrator(
        yara=FakeYara(),
        pg=None,
        router=FakeRouter(),
        worker_pool=pool,
    ) as orchestrator:
        result = orchestrator.analyze_path(path)

    assert result.verdict == "safe"
    assert pool.closed is True


def test_analyze_document_path_can_close_worker_pool(tmp_path: Path):
    path = tmp_path / "safe.txt"
    path.write_text("safe", encoding="utf-8")
    pool = FakeWorkerPool()
    pool.closed = False

    result = analyze_document_path(
        path,
        yara=FakeYara(),
        pg=None,
        router=FakeRouter(),
        worker_pool=pool,
        close_worker_pool=True,
    )

    assert result.verdict == "safe"
    assert pool.closed is True


def test_orchestrator_normalizes_unknown_llm_verdict_to_suspicious(tmp_path: Path):
    path = tmp_path / "doc.txt"
    path.write_text("RISKY-INSTRUCTION", encoding="utf-8")

    class UnknownVerdictPool(FakeWorkerPool):
        def classify_chunks(self, chunks):
            chunk = tuple(chunks)[0]
            return (
                WorkerResult(
                    chunk=chunk,
                    classification=ClassificationResult(
                        verdict="typo-unsafe",
                        confidence=0.6,
                        reasons=("unknown",),
                    ),
                ),
            )

    orchestrator = DocumentOrchestrator(
        yara=FakeYara(),
        pg=None,
        router=FakeRouter(),
        worker_pool=UnknownVerdictPool(),
    )

    result = orchestrator.analyze_path(path)

    assert result.verdict == "suspicious"
    assert result.chunk_results[0].final_verdict == "suspicious"

File: tests/test_orchestration.ts
/**
 * Tests for orchestration.
 * Ported from tests/test_orchestration.py
 */

import { describe, it, expect, vi } from "vitest";
import { writeFileSync, mkdirSync } from "fs";
import { join } from "path";
import {
  DocumentOrchestrator,
  DocumentOrchestratorAsync,
  analyze_document_path,
  VERDICT_SAFE,
  VERDICT_UNSAFE,
  VERDICT_SUSPICIOUS,
  DECISION_HOLD,
  DECISION_REVIEW,
  DECISION_SAFE,
} from "../src/orchestration/orchestrator.js";
import { createClassificationResult } from "../src/classifiers/base.js";
import { createDetectionFinding } from "../src/detection/models.js";
import { createTextChunk } from "../src/ingestion/models.js";
import { TextChunker } from "../src/ingestion/chunking.js";
import { YaraDetector } from "../src/detection/yara.js";
import { CheapRouter } from "../src/detection/detect.js";
import { WorkerResult } from "../src/workers/pool.js";

// ---------------------------------------------------------------------------
// Fakes
// ---------------------------------------------------------------------------

class FakeWorkerPool {
  calls: readonly (readonly createTextChunk[])[] = [];
  closed = false;

  classify_chunks(chunks: readonly createTextChunk[]): readonly WorkerResult[] {
    const chunkList = [...chunks] as readonly createTextChunk[];
    this.calls = [...this.calls, chunkList];

    return chunkList.map((chunk) => ({
      chunk,
      classification: createClassificationResult({
        verdict: chunk.text.includes("RISKY-INSTRUCTION") ? "unsafe" : "safe",
        confidence: 0.9,
        reasons: ["worker"],
        findings: [],
      }),
    }));
  }

  close(): void {
    this.closed = true;
  }
}

class FakeYara {
  constructor(private findings: readonly createDetectionFinding[] = []) {}

  detect(chunk: createTextChunk): readonly createDetectionFinding[] {
    if (!chunk.text.includes("RISKY-INSTRUCTION")) return [];

    const idx = chunk.text.indexOf("RISKY-INSTRUCTION");
    return [
      createDetectionFinding({
        span: "RISKY-INSTRUCTION",
        category: "instruction_override",
        severity: "high",
        reason: "Risk marker for test.",
        rule_id: "risk_marker",
        start_char: chunk.start_char + idx,
        end_char: chunk.start_char + idx + 17,
        source: chunk.source,
        requires_llm_validation: true,
      }),
    ];
  }
}

class FakeRouter {
  constructor(private holdOnRisk = true) {}

  route(yaraFindings: readonly createDetectionFinding[], pgScore: number) {
    if (yaraFindings.length > 0 && this.holdOnRisk) {
      return Object.freeze({
        decision: DECISION_HOLD,
        risk_score: 50.0,
        pg_score: 0.0,
        yara_score: 40.0,
        findings: [],
        reason: "RISKY-INSTRUCTION found.",
      });
    } else if (yaraFindings.length > 0 && !this.holdOnRisk) {
      return Object.freeze({
        decision: DECISION_REVIEW,
        risk_score: 25.0,
        pg_score: 0.0,
        yara_score: 20.0,
        findings: [],
        reason: "RISKY-INSTRUCTION found.",
      });
    }
    return Object.freeze({
      decision: DECISION_SAFE,
      risk_score: 0.0,
      pg_score: 0.0,
      yara_score: 0.0,
      findings: [],
      reason: "No risk signals.",
    });
  }
}

// ---------------------------------------------------------------------------
// Tests
// ---------------------------------------------------------------------------

describe("test_orchestrator_end_to_end_with_index_traceability", () => {
  it("full pipeline with RISKY-INSTRUCTION routes to LLM and returns unsafe", () => {
    // Create temp file via Node.js
    const tmpDir = "/tmp/ozma-test-" + Date.now();
    mkdirSync(tmpDir, { recursive: true });
    const path = join(tmpDir, "doc.txt");
    writeFileSync(
      path,
      "Normal section.\nRISKY-INSTRUCTION appears in this chunk and should route to llm.\nNormal ending.",
      "utf-8",
    );

    const chunker = new TextChunker({ chunkSize: 60, chunkOverlap: 0 });
    const pool = new FakeWorkerPool();
    const orchestrator = new DocumentOrchestrator({
      yara: new FakeYara() as unknown as YaraDetector,
      pg: null,
      router: new FakeRouter() as unknown as CheapRouter,
      workerPool: pool as unknown as import("../src/workers/pool.js").ClassifierWorkerPool,
    });

    const result = orchestrator.analyze_path(path, { chunker });

    expect(result.verdict).toBe(VERDICT_UNSAFE);
    expect(pool.calls.length).toBe(1);
    const routedChunks = pool.calls[0]!;
    expect(routedChunks.length).toBe(1);
    const routedIndex = routedChunks[0]!.metadata["chunk_index"];
    expect(routedIndex).toBeDefined();
  });
});

describe("test_orchestrator_skips_worker_pool_when_no_cheap_findings", () => {
  it("safe document skips worker pool", () => {
    const tmpDir = "/tmp/ozma-test-" + Date.now();
    mkdirSync(tmpDir, { recursive: true });
    const path = join(tmpDir, "safe.txt");
    writeFileSync(path, "No risk markers in this document.", "utf-8");

    const pool = new FakeWorkerPool();
    const orchestrator = new DocumentOrchestrator({
      yara: new FakeYara() as unknown as YaraDetector,
      pg: null,
      router: new FakeRouter() as unknown as CheapRouter,
      workerPool: pool as unknown as import("../src/workers/pool.js").ClassifierWorkerPool,
    });

    const result = orchestrator.analyze_path(path);

    expect(result.verdict).toBe(VERDICT_SAFE);
    expect(pool.calls.length).toBe(0);
  });
});

describe("test_orchestrator_routes_review_chunks_to_layer2", () => {
  it("REVIEW chunks still route to Layer 2", () => {
    const tmpDir = "/tmp/ozma-test-" + Date.now();
    mkdirSync(tmpDir, { recursive: true });
    const path = join(tmpDir, "doc.txt");
    writeFileSync(path, "RISKY-INSTRUCTION", "utf-8");

    const pool = new FakeWorkerPool();
    const orchestrator = new DocumentOrchestrator({
      yara: new FakeYara() as unknown as YaraDetector,
      pg: null,
      router: new FakeRouter(false) as unknown as CheapRouter,
      workerPool: pool as unknown as import("../src/workers/pool.js").ClassifierWorkerPool,
    });

    const result = orchestrator.analyze_path(path);

    expect(pool.calls.length).toBe(1);
  });
});

describe("test_orchestrator_context_manager_closes_worker_pool", () => {
  it("context manager closes pool on exit", () => {
    const tmpDir = "/tmp/ozma-test-" + Date.now();
    mkdirSync(tmpDir, { recursive: true });
    const path = join(tmpDir, "safe.txt");
    writeFileSync(path, "safe", "utf-8");

    const pool = new FakeWorkerPool();
    const orchestrator = new DocumentOrchestrator({
      yara: new FakeYara() as unknown as YaraDetector,
      pg: null,
      router: new FakeRouter() as unknown as CheapRouter,
      workerPool: pool as unknown as import("../src/workers/pool.js").ClassifierWorkerPool,
    });

    orchestrator.close();

    expect(pool.closed).toBe(true);
  });
});

describe("test_orchestrator_normalizes_unknown_llm_verdict_to_suspicious", () => {
  it("unknown verdict normalizes to suspicious", () => {
    const tmpDir = "/tmp/ozma-test-" + Date.now();
    mkdirSync(tmpDir, { recursive: true });
    const path = join(tmpDir, "doc.txt");
    writeFileSync(path, "RISKY-INSTRUCTION", "utf-8");

    class UnknownVerdictPool extends FakeWorkerPool {
      classify_chunks(
        chunks: readonly createTextChunk[],
      ): readonly WorkerResult[] {
        const chunk = chunks[0]!;
        return [
          {
            chunk,
            classification: createClassificationResult({
              verdict: "typo-unsafe",
              confidence: 0.6,
              reasons: ["unknown"],
              findings: [],
            }),
          },
        ];
      }
    }

    const pool = new UnknownVerdictPool();
    const orchestrator = new DocumentOrchestrator({
      yara: new FakeYara() as unknown as YaraDetector,
      pg: null,
      router: new FakeRouter() as unknown as CheapRouter,
      workerPool: pool as unknown as import("../src/workers/pool.js").ClassifierWorkerPool,
    });

    const result = orchestrator.analyze_path(path);

    expect(result.verdict).toBe(VERDICT_SUSPICIOUS);
  });
});
